node-api/backend/validator.go

```diff
@@ -21,80 +21,224 @@
+// given slot. It will then filter them by the provided ids and statuses.
+func (b Backend) FilteredValidators(
+	slot math.Slot, ids []string, statuses []string,
+) ([]*beacontypes.ValidatorData, error) {
+	st, _, err := b.stateFromSlot(slot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get state from slot %d", slot)
+	}
+
+	validators, err := st.GetValidators()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validators")
+	}
+
+	// Parse all IDs and pubkeys once at the start
+	filters := parseValidatorIDs(ids)
+	epoch := b.cs.SlotToEpoch(slot)
+
+	return filterAndBuildValidatorData(st, validators, filters, epoch, statuses)
+}
+
+// filterAndBuildValidatorData processes all validators and builds their data based on filters
+func filterAndBuildValidatorData(
+	st *statedb.StateDB,
+	validators []*types.Validator,
+	filters *validatorFilters,
+	epoch math.Epoch,
+	statuses []string,
+) ([]*beacontypes.ValidatorData, error) {
+	validatorData := make([]*beacontypes.ValidatorData, 0, len(validators))
+
+	for _, validator := range validators {
+		index, err := st.ValidatorIndexByPubkey(validator.GetPubkey())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get validator index by pubkey")
+		}
+
+		if !matchesFilters(validator, index, filters) {
+			continue
+		}
+
+		data, err := buildValidatorData(st, validator, index, epoch, statuses)
+		if err != nil {
+			return nil, err
+		}
+		if data != nil {
+			validatorData = append(validatorData, data)
+		}
+	}
+
+	return validatorData, nil
+}
+
+// matchesFilters checks if a validator matches the filters.
+func matchesFilters(validator *types.Validator, index math.U64, filters *validatorFilters) bool {
+	// If no filters, accept all validators
+	if len(filters.numericIDs) == 0 && len(filters.pubkeys) == 0 {
+		return true
+	}
+
+	// Check numeric IDs
+	if len(filters.numericIDs) > 0 && matchesIndex(index, filters.numericIDs) {
+		return true
+	}
+
+	// Check pubkeys
+	if len(filters.pubkeys) > 0 && matchesPubkey(validator, filters.pubkeys) {
+		return true
+	}
+
+	return false
+}
+
+func matchesPubkey(validator *types.Validator, parsedPubkeys []crypto.BLSPubkey) bool {
+	validatorPubkey := validator.GetPubkey()
+	return slices.Contains(parsedPubkeys, validatorPubkey)
+}
+
+func matchesIndex(index math.U64, ids []uint64) bool {
+	return slices.Contains(ids, index.Unwrap())
+}
+
+func matchesStatusFilter(status string, statuses []string) bool {
+	return len(statuses) == 0 || slices.Contains(statuses, status)
+}
+
+func buildValidatorData(
+	st *statedb.StateDB,
+	validator *types.Validator,
+	index math.U64,
+	epoch math.Epoch,
+	statuses []string,
+) (*beacontypes.ValidatorData, error) {
+	status, err := validator.Status(epoch)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validator status")
+	}
+
+	if !matchesStatusFilter(status, statuses) {
+		//nolint:nilnil // no data to return
```

Comment:
Would agree that rather than returning nil, nil, we should return a defined error like ErrNoValidatorData and then explicitly catch this error case in the calling function.

Will remove the nolinting of nil,nil as well

Reply:
makes sense, returned an error and then handled it in `filterAndBuildValidatorData` so that when no status matches, error is not returned (as per spec).